File: stack/stack.py
#! /usr/bin/python
# -*- coding:utf-8 -*-


class Node:
    def __init__(self, data):
        self.data = data
        self.prevNode = None
        self.nextNode = None

class Stack:
    def __init__(self):
        self.headNode = None
        self.tailNode = None
        self.elementCount = 0

    def GetSize(self):
        return self.elementCount

    def PrintData(self):
        if self.headNode is None:
            print("출력할 데이터가 없습니다.")
            return

        count = 0
        currentNode = self.headNode
        while currentNode:
            count += 1
            print(count, '번째 데이터 : ', currentNode.data)
            currentNode = currentNode.nextNode
            if currentNode == self.headNode:
                return

    def Append(self, data):
        appendNode = Node(data)
        if self.headNode is None:
            self.headNode = appendNode
            self.tailNode = self.headNode
        else:
            currentNode = self.tailNode
            currentNode.nextNode = appendNode
            appendNode.prevNode = currentNode
            self.tailNode = appendNode
        self.tailNode.nextNode = self.headNode
        self.headNode.prevNode = self.tailNode
        self.elementCount += 1

    def Pop(self):
        if self.headNode is None:
            print("데이터가 없습니다. pop실패")
            return
        popNode = self.tailNode
        data = popNode.data
        self.tailNode = popNode.prevNode
        self.tailNode.nextNode = self.headNode
        self.headNode.prevNode = self.tailNode
        popNode = None
        self.elementCount -= 1
        return data


if __name__ == '__main__':
    stack = Stack()
    stack.PrintData()
    print('스택 사이즈 : ', stack.GetSize())

    stack.Append(100)
    stack.PrintData()
    print('스택 사이즈 : ', stack.GetSize())

    for i in range(1, 10):
        stack.Append(i)
    stack.PrintData()
    print('스택 사이즈 : ', stack.GetSize())

    print('Pop 데이터 : ', stack.Pop())
    print('Pop 데이터 : ', stack.Pop())
    stack.PrintData()
    print('스택 사이즈 : ', stack.GetSize())

File: stack/stack.cpp
#include <iostream>

struct Node
{
    Node * PrevNode;
    Node * NextNode;
    int Data;
    Node(int InData = 0)
    {
        PrevNode = nullptr;
        NextNode = nullptr;
        Data = InData;
    }
};

void SafeDelete(Node * InNode)
{
    if(InNode)
    {
        delete InNode;
        InNode = nullptr;
    }
}

class Stack
{
    private:
        Node * HeadNode;
        Node * TailNode;
        int ElementCount;

    public:
        Stack();
        ~Stack();
        bool IsEmpty();
        int GetSize();
        bool PushData(int InData);
        int PopData();
        void PrintData();
};

Stack::Stack()
{
    HeadNode = nullptr;
    TailNode = nullptr;
    ElementCount = 0;
}

Stack::~Stack()
{
    if(TailNode)
    {
        Node * TempNode = TailNode->PrevNode;
        while(TempNode)
        {
            Node * DeleteNode = TempNode;
            TempNode = TempNode->PrevNode;
            SafeDelete(DeleteNode);
            ElementCount--;
        }
       
        TailNode = nullptr;
    }

    HeadNode = nullptr;
}

bool Stack::IsEmpty()
{
    return ElementCount == 0;
}

int Stack::GetSize()
{
    return ElementCount;
}

bool Stack::PushData(int InData)
{
    if(HeadNode == nullptr)
    {
        Node * PushNode = new Node(InData);
        HeadNode = PushNode;
        TailNode = HeadNode;
    }
    else
    {
        Node * PushNode = new Node(InData);
        Node * TempNode = this->TailNode;
        TempNode->NextNode = PushNode;
        PushNode->PrevNode = TempNode;
        this->TailNode = PushNode;
    }

    ElementCount++;

    return true;
}

int Stack::PopData()
{
    if(ElementCount == 0)
    {
        std::cout << "스택에 데이터가 없습니다." << std::endl;
        return -1;
    }

    int Data = 0;
    Node * PopNode = this->TailNode;
    Data = PopNode->Data;
    PopNode->PrevNode->NextNode = nullptr;
    this->TailNode = PopNode->PrevNode;
    SafeDelete(PopNode); 
    ElementCount--;

    return Data;
}

void Stack::PrintData()
{
    int Count = 0;
    Node * TempNode = HeadNode;
    while(TempNode)
    {
        std::cout << Count + 1 << "번째 데이터 : " << TempNode->Data << std::endl;
        TempNode = TempNode->NextNode;
    }
}

int main()
{
    Stack StackList;

    std::cout << "스택 사이즈 : " << StackList.GetSize() << std::endl;
    for(int i=0; i<15; ++i)
    {
        StackList.PushData(i);
    }

    StackList.PrintData();
    std::cout << "스택 사이즈 : " << StackList.GetSize() << std::endl;
    
    
    for(int j=0; j<3; ++j)
    {
        std::cout << StackList.PopData() << " 데이터 팝업" << std::endl;
    }

    StackList.PrintData();
    std::cout << "스택 사이즈 : " << StackList.GetSize() << std::endl;
    
}
